virtcontainers/qemu.go

```diff
@@ -1237,6 +1261,72 @@ func (q *qemu) disconnect() {
 	q.qmpShutdown()
 }
 
+// resizeMemory get a request to update the VM memory to reqMemMB
+// Memory update is managed with two approaches
+// Add memory to VM:
+// When memory is required to be added we hotplug memory
+// Remove Memory from VM/ Return memory to host.
+//
+// Memory unplug can be slow and it cannot be guaranteed.
+// Additionally, the unplug has not small granularly it has to be
+// the memory to remove has to be at least the size of one slot.
+// To return memory back we are resizing the VM memory balloon.
+// A longer term solution is evaluate solutions like virtio-mem
+func (q *qemu) resizeMemory(reqMemMB uint32, memoryBlockSizeMB uint32) (uint32, error) {
+
+	currentMemory := q.config.MemorySize + uint32(q.state.HotpluggedMemory)
+	err := q.qmpSetup()
+	if err != nil {
+		return 0, err
+	}
+	switch {
+	case currentMemory < reqMemMB:
+		//hotplug
+		addMemMB := reqMemMB - currentMemory
+		memHotplugMB, err := calcHotplugMemMiBSize(addMemMB, memoryBlockSizeMB)
+		if err != nil {
+			return currentMemory, err
+		}
+
+		addMemDevice := &memoryDevice{
+			sizeMB: int(memHotplugMB),
+		}
+		data, err := q.hotplugAddDevice(addMemDevice, memoryDev)
+		if err != nil {
+			return currentMemory, err
+		}
+		memoryAdded, ok := data.(int)
+		if !ok {
+			return currentMemory, fmt.Errorf("Could not get the memory added, got %+v", data)
+		}
+		currentMemory += uint32(memoryAdded)
+	case currentMemory > reqMemMB:
+		//hotunplug
+		addMemMB := currentMemory - reqMemMB
+		memHotunplugMB, err := calcHotplugMemMiBSize(addMemMB, memoryBlockSizeMB)
+		if err != nil {
+			return currentMemory, err
+		}
+
+		addMemDevice := &memoryDevice{
+			sizeMB: int(memHotunplugMB),
+		}
+		data, err := q.hotplugRemoveDevice(addMemDevice, memoryDev)
```

Comment:
Since we don't support hotunplug memory right now, I would prefer we do not subtract the memory from total memory and we are correct currentMemory at least. A FIXME might be better here if we do not want to fail `hotplugRemoveDevice` as of now. I understand you kept the old code as is, so maybe a follow github issue to fix the behavior might be better. Otherwise if someone does `update` frequently can end up with a vm with much more memory than actually being used.

Reply:
There is nothing to worry about here since the code path from `hotplugRemoveDevice()` leads to `hotplugDevice()` which leads to `hotplugMemory()`. And here is the code from `hotplugMemory`:
```go
        if memDev.sizeMB < 0 {
		return 0, fmt.Errorf("cannot hotplug negative size (%d) memory", memDev.sizeMB)
	}

	// We do not support memory hot unplug.
	if op == removeDevice {
		// Dont fail for now, until we fix it.
		// We return that we only unplugged 0
		q.Logger().Warn("hot-remove VM memory not supported")
		return 0, nil
	}

	err := q.qmpSetup()
	if err != nil {
		return 0, err
	}
```

So the data returned will be `0` in case of removal, hence the amount of memory returned is gonna be the `currentMemory`.
Am I missing something?

Reply:
I added an extra validation to verify this is always 0, I hope this make clear that we expect 0 at this moment.